src/signaling/mcu_proxy.go

```diff
@@ -0,0 +1,989 @@
+		if err != nil {
+			return nil, err
+		}
+
+		connections = append(connections, conn)
+	}
+	if len(connections) == 0 {
+		return nil, fmt.Errorf("No MCU proxy connections configured")
+	}
+
+	mcu.setConnections(connections)
+	return mcu, nil
+}
+
+func (m *mcuProxy) setConnections(connections []*mcuProxyConnection) {
+	m.connections.Store(connections)
+}
+
+func (m *mcuProxy) getConnections() []*mcuProxyConnection {
+	return m.connections.Load().([]*mcuProxyConnection)
+}
+
+func (m *mcuProxy) Start() error {
+	for _, c := range m.getConnections() {
+		if err := c.start(); err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+func (m *mcuProxy) Stop() {
+	for _, c := range m.getConnections() {
+		ctx, cancel := context.WithTimeout(context.Background(), closeTimeout)
+		defer cancel()
+		c.stop(ctx)
+	}
+}
+
+func (m *mcuProxy) SetOnConnected(f func()) {
+	// Not supported.
+}
+
+func (m *mcuProxy) SetOnDisconnected(f func()) {
+	// Not supported.
+}
+
+type mcuProxyStats struct {
+	Publishers int64                               `json:"publishers"`
+	Clients    int64                               `json:"clients"`
+	Details    map[string]*mcuProxyConnectionStats `json:"details"`
+}
+
+func (m *mcuProxy) GetStats() interface{} {
+	details := make(map[string]*mcuProxyConnectionStats)
+	result := &mcuProxyStats{
+		Details: details,
+	}
+	for _, conn := range m.getConnections() {
+		stats := conn.GetStats()
+		result.Publishers += stats.Publishers
+		result.Clients += stats.Clients
+		details[stats.Url] = stats
+	}
+	return result
+}
+
+type mcuProxyConnectionsList []*mcuProxyConnection
+
+func (l mcuProxyConnectionsList) Len() int {
+	return len(l)
+}
+
+func (l mcuProxyConnectionsList) Less(i, j int) bool {
+	return l[i].Load() < l[j].Load()
+}
+
+func (l mcuProxyConnectionsList) Swap(i, j int) {
+	l[i], l[j] = l[j], l[i]
+}
+
+func (l mcuProxyConnectionsList) Sort() {
+	sort.Sort(l)
+}
+
+func (m *mcuProxy) getSortedConnections() []*mcuProxyConnection {
+	connections := m.getConnections()
+	if len(connections) < 2 {
+		return connections
+	}
+
+	// Connections are re-sorted every <connectionSortRequests> requests or
+	// every <connectionSortInterval>.
+	now := time.Now().UnixNano()
+	if atomic.AddInt64(&m.connRequests, 1)%connectionSortRequests == 0 || atomic.LoadInt64(&m.nextSort) <= now {
+		atomic.StoreInt64(&m.nextSort, now+int64(connectionSortInterval))
+
+		sorted := make(mcuProxyConnectionsList, len(connections))
+		copy(sorted, connections)
+
+		sorted.Sort()
+
+		m.setConnections(sorted)
+		connections = sorted
+	}
+
+	return connections
+}
+
+func (m *mcuProxy) removePublisher(publisher *mcuProxyPublisher) {
+	m.mu.Lock()
+	defer m.mu.Unlock()
+
+	delete(m.publishers, publisher.Id())
+}
+
+func (m *mcuProxy) NewPublisher(ctx context.Context, listener McuListener, id string, streamType string) (McuPublisher, error) {
+	connections := m.getSortedConnections()
+	for _, conn := range connections {
+		if conn.IsShutdownScheduled() {
```

Comment:
This will be true if the proxy is started again after being shut down. Maybe the fields need to be reinitialized in the signaling server when connecting again with a proxy? Or a new object was expected to have been used but a previous one was reused?

Reply:
Right, I changed this to reset the flag on reconnections (will be set again after the hello has been processed if the proxy is still scheduled to shutdown).